File: demo_android/EmptyLoading/library/src/main/java/better/android/library/empty/LoadingProxy.java
package better.android.library.empty;

import android.util.Log;
import android.view.View;
import android.view.ViewGroup;

/**
 * 替换对应的ContentView
 * Created by better on 2018/3/17 22:40.
 */

public class LoadingProxy implements EmptyViewRefreshProxy {

    private View mInitView;
    private ViewGroup mInitParent;
    private int mInitIndex;

    private EmptyViewRefreshProxy mRefreshProxy;

    public LoadingProxy setRefreshProxy(EmptyViewRefreshProxy refreshProxy) {
        mRefreshProxy = refreshProxy;
        return this;
    }

    public LoadingProxy(View view) {

        mRefreshProxy = new EmptyViewProxy(view.getContext());

        mInitView = view;
        if (mInitView.getParent() != null) {
            mInitParent = (ViewGroup) mInitView.getParent();
        } else {
            mInitParent = mInitView.getRootView().findViewById(android.R.id.content);
        }

        for (int i = 0; i < mInitParent.getChildCount(); i++) {
            if (mInitParent.getChildAt(i).getId() == mInitView.getId()) {
                mInitIndex = i;
            }
        }

    }

    private void log(String msg) {
        if (BuildConfig.DEBUG) {
            Log.d(this.getClass().getSimpleName(), msg);
        }
    }

    private void removeLoadingView() {
        if (mInitParent.getChildAt(mInitIndex) == mRefreshProxy.getProxyView()) {
            mInitParent.removeViewAt(mInitIndex);
            mInitParent.addView(mInitView, mInitIndex, mInitView.getLayoutParams());
            log("removeLoadingView");
        }
//        for (int i = 0; i < mInitParent.getChildCount(); i++) {
//            if (mInitParent.getChildAt(i) == mRefreshProxy.getProxyView()) {
//                mInitParent.removeViewAt(i);
//                mInitParent.addView(mInitView, i, mInitView.getLayoutParams());
//                log("removeLoadingView");
//            }
//        }

    }

    private void prepareLoadingView() {
        if (mInitParent.getChildAt(mInitIndex) == mInitView) {
            mInitParent.removeViewAt(mInitIndex);
            mInitParent.addView(mRefreshProxy.getProxyView(), mInitIndex, mInitView.getLayoutParams());
        }
    }

    @Override
    public void displayLoading() {
        prepareLoadingView();
        if (null != mRefreshProxy) {
            mRefreshProxy.displayLoading();
        }
    }

    @Override
    public void displayLoading(int msgId) {
        prepareLoadingView();
        if (null != mRefreshProxy) {
            mRefreshProxy.displayLoading(msgId);
        }
    }

    @Override
    public void displayLoading(String msg) {
        prepareLoadingView();
        if (null != mRefreshProxy) {
            mRefreshProxy.displayLoading(msg);
        }
    }

    @Override
    public void displayMessage() {
        prepareLoadingView();
        if (null != mRefreshProxy) {
            mRefreshProxy.displayMessage();
        }
    }

    @Override
    public void displayMessage(int msg) {
        prepareLoadingView();
        if (null != mRefreshProxy) {
            mRefreshProxy.displayMessage(msg);
        }
    }

    @Override
    public void displayMessage(String msg) {
        prepareLoadingView();
        if (null != mRefreshProxy) {
            mRefreshProxy.displayMessage(msg);
        }
    }


    @Override
    public void displayRetry() {
        prepareLoadingView();
        if (null != mRefreshProxy) {
            mRefreshProxy.displayRetry();
        }
    }

    @Override
    public void displayRetry(int msgId) {
        prepareLoadingView();
        if (null != mRefreshProxy) {
            mRefreshProxy.displayRetry(msgId);
        }
    }

    @Override
    public void displayRetry(String msg) {
        prepareLoadingView();
        if (null != mRefreshProxy) {
            mRefreshProxy.displayRetry(msg);
        }
    }

    @Override
    public void disappear() {
        removeLoadingView();
        if (null != mRefreshProxy) {
            mRefreshProxy.disappear();
        }
    }

    @Override
    public View getProxyView() {
        if (null != mRefreshProxy) {
            return mRefreshProxy.getProxyView();
        }
        return null;
    }

    @Override
    public void setOnRetryClickListener(OnLrpRetryClickListener listener) {
        if (null != mRefreshProxy) {
            mRefreshProxy.setOnRetryClickListener(listener);
        }
    }
}

File: demo_android/EmptyLoading2/libraryEmpty/src/main/java/com/better/library/empty/LoadingProxy.kt
package com.better.library.empty

import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import androidx.core.view.get

class LoadingProxy(private val wrapper: FrameLayout) : EmptyViewRefreshProxy {
    companion object {
        fun wrap(view: View): LoadingProxy {
            val wrapper = FrameLayout(view.context)
            if (null == view.parent) {
                wrapper.layoutParams = view.layoutParams
            } else {
                val parent = view.parent as ViewGroup
                var index = 0
                for (i in (0 until parent.childCount)) {
                    if (parent[i] == view) {
                        index = i
                        break
                    }
                }
                parent.removeView(view)
                parent.addView(wrapper, index, view.layoutParams)
            }
            wrapper.addView(view)
            return LoadingProxy(wrapper)
        }
    }

    private var emptyProxy = EmptyViewProxyImpl(wrapper.context)
        set(value) {
            val params = field.proxyView.layoutParams as ViewGroup.LayoutParams
            wrapper.removeView(field.proxyView)
            field = value
            wrapper.addView(field.proxyView, params)
        }

    override val proxyView: View
        get() = wrapper

    init {
        val params = FrameLayout.LayoutParams(
            FrameLayout.LayoutParams.MATCH_PARENT,
            FrameLayout.LayoutParams.MATCH_PARENT
        )
        wrapper.addView(emptyProxy.proxyView, params)
    }

    override fun displayLoading(msg: Int) {
        emptyProxy.displayLoading(msg)
    }

    /**
     * 显示请求等待。
     *
     * @param msg 等待提示消息。
     */
    override fun displayLoading(msg: String) {
        emptyProxy.displayLoading(msg)
    }

    override fun displayMessage(msg: Int) {
        emptyProxy.displayMessage(msg)
    }

    /**
     * 请求成功 当列表为空的时候展示的信息
     * Create By better on 2017/4/2 07:14.
     */
    override fun displayMessage(msg: String) {
        emptyProxy.displayMessage(msg)
    }

    override fun displayRetry(msg: Int) {
        emptyProxy.displayRetry(msg)
    }

    /**
     * 请求失败 显示请求重试。
     *
     * @param msg 重试提示消息：失败、无数据。
     */
    override fun displayRetry(msg: String) {
        emptyProxy.displayRetry(msg)
    }

    /**
     * 直接消失emptyView
     */
    override fun disappear() {
        emptyProxy.disappear()
    }

    override fun setOnRetryClickListener(listener: OnLrpRetryClickListener) {
        emptyProxy.setOnRetryClickListener(listener)
    }

}